presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java

```diff
@@ -113,6 +119,44 @@ protected SymbolStatsEstimate visitLiteral(Literal node, Void context)
             return estimate.build();
         }
 
+        @Override
+        protected SymbolStatsEstimate visitFunctionCall(FunctionCall node, Void context)
+        {
+            Map<NodeRef<Expression>, Type> expressionTypes = getExpressionTypes(session, node, types);
+            ExpressionInterpreter interpreter = ExpressionInterpreter.expressionOptimizer(node, metadata, session, expressionTypes);
```

Comment:
We do, but magic literal functions don't get optimized (can't be if we want to preserve expression type)

Reply:
But the code that invokes `ExpressionInterpreter` looks exactly the same. Why the code in the `FilterStatsCalculator` doesn't inline the magic literal? (https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/cost/FilterStatsCalculator.java#L101)

Reply:
Optimized expression needs to be represented as Expression nodes again. Please take a look at the code that does that https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/LiteralEncoder.java#L183
For instance varbinary constants are always wrapped in magic litear function call.

Reply:
Nevermind, got it.

So, basically if this is a filter expression, say `<column1> = magic_literal_function("1234566")`, it will leave the `magic_literal_function` untouched, as it cannot interpret the expression as a whole. That's why we need to run optimizer when `magic_literal_function` is on top of the tree.